### PR TITLE
Sugar tests

### PR DESCRIPTION
**Pull request by konsumlamm**
Followup to #16802.

* remove unnecessary example in `sugar` (the example also works without `capture` and I decided to not replace it with another example, since situations where you need `capture` in `mapIt` are rare)
* add more tests (for `=>`, `->`, `dump` and `capture`)
* put tests in `block`s

### CODE REVIEW

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [inline tests/stdlib/tsugar.nim]: group lexically, eg:
```nim
block: # `=>`
  block: # (in case you need sub-blocks)
    ...
  block:
     ...
block: # `->`
  ...
```
makes it easier to follow and search for tests for a given proc

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [inline tests/stdlib/tsugar.nim]: nit: i usually prefer f1, f2 etc insetead of f, g when there's a lot of such procs to test
```suggestion
  let f2 = () => ()
  doAssert f2() == () # empty tuple, not void

  const f3 = () => ()
  doAssert f3() == ()

  # const f4 = () => void # xxx we should have a way to generate a void-returning proc 
```

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [inline tests/stdlib/tsugar.nim]: future PR welcome to do the following: https://github.com/timotheecour/Nim/issues/534#issuecomment-769565033
(or in this PR if you can)

File: tests/stdlib/tsugar.nim
Comment (konsumlamm) [reply tests/stdlib/tsugar.nim]: Do you mean for all procs? E.g. for `collect`, there are quite a lot of tests.

Comment (timotheecour) [reply tests/stdlib/tsugar.nim]: at least the ones you're adding/editing would be nice

File: tests/stdlib/tsugar.nim
Comment (konsumlamm) [reply tests/stdlib/tsugar.nim]: That actually made me find a bug (#12595) with using `{k: v}` in templates.

File: tests/stdlib/tsugar.nim
Comment (konsumlamm) [reply tests/stdlib/tsugar.nim]: What's the point of testing `() => ()`?

Also, not sure what you mean by the last comment, returning `void` seems to work.

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [inline tests/stdlib/tsugar.nim:45]: add this:
```nim
      let f5 = () => (discard) # simplest proc that returns void
      f5()
```
and also add this to runnableExamples in sugar.=> as it's not obvious how to write it

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [inline tests/stdlib/tsugar.nim:44]: can you also add a proc litteral that takes a pragma? (pick one, maybe `{.inline.}`?) ditto with runnableExamples to show its syntax, however obvious it may be

ditto for `->`

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [reply tests/stdlib/tsugar.nim]: indeed, following up in https://github.com/nim-lang/Nim/pull/16820#discussion_r567195878

File: tests/stdlib/tsugar.nim
Comment (konsumlamm) [inline tests/stdlib/tsugar.nim]: When #16874 gets merged, this can be put in the `main` template

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [inline tests/stdlib/tsugar.nim]: you can use `let x {.inject.} = 10` to prevent that

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [inline tests/stdlib/tsugar.nim]: you can avoid code duplication as follows:

```nim
proc mainProc() =
  # symbols in templates are gensym'd
  block: # dump
    let
      x = 10
      y = 20
    dump(x + y) # x + y = 30

  # TODO: move to `main` and `when false` above pending bug #12595
  block: # collect pairs
    let data = @["bird", "word"]
    doAssert collect(initTable(2), for (i, d) in data.pairs: {i: d}) ==
      {0: "bird", 1: "word"}.toTable
    doAssert collect(for (i, d) in pairs(data): {i: d}) ==
      {1: "word", 0: "bird"}.toTable

static: mainProc()
mainProc()
```

File: tests/stdlib/tsugar.nim
Comment (konsumlamm) [reply tests/stdlib/tsugar.nim]: Good idea, though that made me rediscover bug #10609 (apparently using `{.inject.}` once applies to the whole template).

Comment (timotheecour) [reply tests/stdlib/tsugar.nim]: note that https://github.com/nim-lang/Nim/pull/16820#pullrequestreview-579800457 obsoletes this suggestion

File: tests/stdlib/tsugar.nim
Comment (timotheecour) [inline tests/stdlib/tsugar.nim]: make sure to update this now that your fix to https://github.com/nim-lang/Nim/issues/12595 got merged :)

File: tests/stdlib/tsugar.nim
Comment (konsumlamm) [reply tests/stdlib/tsugar.nim:45]: So for some reason, `runnableExamples` transforms
```nim
let f = () => (discard)
```
to
```nim
let f = () =>
  discard
```
which doesn't compile... (but that doesn't happen when using for example `echo "test"` instead of `discard`)

Comment (timotheecour) [reply tests/stdlib/tsugar.nim:45]: i see, you're hitting https://github.com/nim-lang/Nim/issues/13491

File: lib/pure/sugar.nim
Comment (timotheecour) [inline lib/pure/sugar.nim]:
```suggestion
    # proc passOne(f: proc (x: int): int {.noSideEffect.}): int = f(1)
```

File: tests/stdlib/tsugar.nim
Comment (konsumlamm) [reply tests/stdlib/tsugar.nim:45]: How should i proceed on this one?

Comment (timotheecour) [reply tests/stdlib/tsugar.nim:45]: * add it in runnableExamples but commented out
* add a comment in code: # xxx pending #13491: uncomment runnableExamples
* add a test in tsugar

**Review by timotheecour — APPROVED** [review]
LGTM; excellent.

future work can:
* maybe re-add to tsugar an improved version of the `mapIt` example that was removed (one where it would actually be needed); 
* add tests for `(x: int) -> int` (there are currently to tests for that, only  `int -> int`) and in particular ones showing you can pass a proc with differently named params to it

**Comment by timotheecour** [comment]
still LGTM but in a followup PR (or feel free to do it here), I'll improve as follows:

* test in js too with:
```
discard """
  targets: "c js"
  output: '''
x + y = 30
'''
"""
```

* this requires only 1 disabled test for js: (please file a bug so we can link to it in this test `# pending bug #xyz`)
```nim
  block: # capture
    var closure1: () -> int
    when not defined js: # xxx gives: Error: cannot evaluate at compile time: i`gensym0
      for i in 0 .. 10:
        if i == 5:
          capture i:
            closure1 = () => i
      doAssert closure1() == 5

      var closure2: () -> (int, int)
      for i in 0 .. 10:
        for j in 0 .. 10:
          if i == 5 and j == 3:
            capture i, j:
              closure2 = () => (i, j)
      doAssert closure2() == (5, 3)
```

* and also put all in the main template, by using `inject` in just those places:

```nim

  block: # dump
    # symbols in templates are gensym'd
    let
      x {.inject.} = 10
      y {.inject.} = 20
    dump(x + y) # x + y = 30

  block: # dumpToString
    template square(x): untyped = x * x
    let x {.inject.} = 10
    doAssert dumpToString(square(x)) == "square(x): x * x = 100"
    let s = dumpToString(doAssert 1+1 == 2)
    doAssert "failedAssertImpl" in s
    let s2 = dumpToString:
      doAssertRaises(AssertionDefect): doAssert false
    doAssert "except AssertionDefect" in s2

static: main()
main()
```
(I'm not seeing the bug you mentioned here: https://github.com/nim-lang/Nim/pull/16820#discussion_r567321609)

EDIT => see PR https://github.com/nim-lang/Nim/pull/16919

**Comment by konsumlamm** [comment]
> test in js too

I'll leave that to another PR (I currently don't even have JS installed on my PC). What about other targets (C++, ObjC)? Also, shouldn't the default be "all targets"?

> (I'm not seeing the bug you mentioned here: [#16820 (comment)](https://github.com/nim-lang/Nim/pull/16820#discussion_r567321609))

When using `{.inject.}`, the identifier becomes injected for the whole template (i.e. the injection is not scoped), excluding declarations. So for example the following (from one of the tests for `collect`) will break:
```nim
let x = collect: # here the x is gensym'd
  for d in data.items:
    when d is int: "word"
    else: d
doAssert x == @["bird", "word"] # here the x is injected, due to the mentioned bug
```
This doesn't occur in your example, since every `x` is gensym'd there.

**Comment by timotheecour** [comment]
> I'll leave that to another PR

sounds good, this is LGTM already, just needs another LGTM from 1 other person...

> (I currently don't even have JS installed on my PC).

time for https://nodejs.org/en/download/? this will help you in other PRs

> What about other targets (C++, ObjC)?

sad but there's currently only 1 test(!!) for objc: `tests/osproc/passenv.nim:4:3:  targets: "c cpp objc"`; this can be improved in future work; although swift is more common nowadays

> Also, shouldn't the default be "all targets"?

`cpp` is usually on a need by need basis, eg for really important tests or things that have failed in C++ before or things likely to fail in C++ or c++ specific stuff; the only problem with testing all targets is CI time to run those. 

- [ ] proposal: `targets: "default js"`
That said, CI runs `cpp`  in CI on 1 osx instance (see `NIM_COMPILE_TO_CPP`), but it's not bullet proof (eg when `targets: "c js"` is specified, it won't run cpp; PR welcome to change this to `targets: "default js"`)

(EDIT => https://github.com/nim-lang/RFCs/issues/334)

> So for example the following (from one of the tests for collect) will break:

not sure what you mean since it worked for me, but we can discuss this in next PR

**Review by ringabout — APPROVED** [review]

**Pull request merged**
